src/lightning/fabric/utilities/throughput_monitor.py

```diff
@@ -0,0 +1,367 @@
+    +--------------------------+---------------------------------------------------------------------------------------+
+    | `device/flops_per_sec`   | `flops_per_sec` divided by world size.                                                |
+    +--------------------------+---------------------------------------------------------------------------------------+
+    | `device/mfu`             | `device/flops_per_sec` divided by world size.                                         |
+    +--------------------------+---------------------------------------------------------------------------------------+
+    | `time`                   | Total elapsed time                                                                    |
+    +--------------------------+---------------------------------------------------------------------------------------+
+
+    Notes:
+        - The implementation assumes that devices FLOPs are all the same as it normalizes by the world size and only
+            takes a single ``flops_available`` value.
+        - items_per_sec, flops_per_sec and MFU do not account for padding if present. We suggest using
+            samples_per_sec or batches_per_sec to measure throughput under this circumstance.
+
+    Args:
+        flops_available: Number of theoretical flops available for a single device.
+        world_size: Number of devices available across hosts. Global metrics are not included if the world size is 1.
+        window_size: Number of batches to use for a rolling average.
+        time_unit: Time unit to use for normalization.
+        separator: Key separator to use when creating per-device and global metrics.
+
+    """
+
+    def __init__(
+        self,
+        flops_available: Optional[float] = None,
+        world_size: int = 1,
+        window_size: int = 100,
+        time_unit: Literal["seconds", "minutes", "hours", "days"] = "hours",
+        separator: str = "/",
+    ) -> None:
+        self.flops_available = flops_available
+        self.separator = separator
+        assert world_size > 0
+        self.world_size = world_size
+
+        # throughput is computed over a window of batches
+        self._samples: Deque[int] = deque(maxlen=window_size + 1)
+        self._time: Deque[float] = deque(maxlen=window_size + 1)
+        self._lengths: Deque[int] = deque(maxlen=window_size + 1)
+        self._flops: Deque[int] = deque(maxlen=window_size + 1)
+
+        if time_unit == "seconds":
+            self._divider = 1
+        elif time_unit == "minutes":
+            self._divider = 60
+        elif time_unit == "hours":
+            self._divider = 60 * 60
+        elif time_unit == "days":
+            self._divider = 60 * 60 * 24
+        else:
+            raise ValueError(
+                f'Invalid time_unit: {time_unit}. Must be one of "seconds", "minutes", "hours", or "days".'
+            )
+
+    def compute(
```

Comment:
I implemented some differences to the GPT implementation. This class now has a single `compute` method, meaning that it doesn't know about the training stage at all (train vs validation).

So in Fabric, one can choose to only call it during training, or during training and validation, or during all stages.